### PR TITLE
Flekschas/viewconf editor

### PR DESCRIPTION
**Pull request by flekschas**
## Description

> What was changed in this pull request?

Added a dialog for directly editing the view config in the browser

> Why is it necessary?

It's handy for changing options quickly that do not have any associated UI element.

> How does it look like?

viewconf-editor

> But I like them keyboard shortcuts

We hear you!
- To update the viewconf hit `⌘+S` or `CTRL+S`
- To temporarily hide the editor click and hold `ALT` for 1 second. As soon as you release it the editor will be visible again
- To cancel and revert already saved changed hit `ESC`
- To save and close the editor hit `⌘+Enter` or `CTRL+Enter`

## Checklist

- [x] Unit tests added or updated
- ~[ ] Documentation added or updated~
- ~[ ] Example added or updated~
- [x] Screenshot for visual changes (e.g. new tracks)
- [x] Updated CHANGELOG.md

### CODE REVIEW

File: app/scripts/ViewConfigEditor.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    this.pubSubs.push(
+      this.props.pubSub.subscribe('keyup', this.handleKeyUpBound),
+    );
+  }
+
+  async componentDidMount() {
+    if (this.editor) {
+      this.editor._input.focus();
+      this.editor._input.setSelectionRange(0, 0);
+      await timeout(0);
```
Comment (pkerpedjiev) [inline app/scripts/ViewConfigEditor.js]: What's this for?

Comment (flekschas) [reply app/scripts/ViewConfigEditor.js]: This autofocuses the textarea and scrolls to the top. Apparently when the content in the textarea is large and the textarea is focused Chrome will automatically scroll to the end of it, which is a little confusing when you first load the view config.

**Comment by pkerpedjiev** [comment]
This is awesome! Especially the keyboard shortcuts. Would you mind creating a docs page showing this editor and including the information about the shortcuts?

**Comment by pkerpedjiev** [comment]
Also, just out of curiosity, how are the shortcuts implemented?

**Comment by flekschas** [comment]
Sure I can add a page. Also, the shortcuts are displayed on the button to make it easy for people to discover them.

Take a look at `handleKeydown()` and `handleKeyup()` in `ViewConfigEditor`. The implementation is pretty straight forward. Below is one of the functions as an example. All you have to do is check that the `key` property matches and one of the modifier keys is active (if that is important)

```js
  handleKeyDown(event) {
    if (event.altKey) {
      this.willHide = setTimeout(this.hideBound, 1000);
    }
    if (event.key === 's' && (event.ctrlKey || event.metaKey)) {
      event.preventDefault();
      this.props.onChange(this.state.code);
    }
    if (event.key === 'Enter' && (event.ctrlKey || event.metaKey)) {
      event.preventDefault();
      this.props.onChange(this.state.code);
      this.props.modal.close();
    }
  }
```

The only other important thing is to call `event.preventDefault();` to avoid duplicated actions, e.g., when hitting `CMD+S`.

**Comment by flekschas** [comment]
@pkerpedjiev Docs are in. See https://github.com/higlass/higlass/pull/649/commits/40ac406ba4ffb6f10981cb4a9e23ca79b00e5918

**Pull request merged**